src/Tmds.MDns/MdnsSocket.cs

```diff
@@ -0,0 +1,142 @@
+using System.Linq;
+using System.Net;
+using System.Net.NetworkInformation;
+using System.Net.Sockets;
+using System.Threading;
+using System.Runtime.InteropServices;
+
+namespace Tmds.MDns
+{
+    class MdnsSocket : IDisposable
+    {
+        public static readonly IPEndPoint IPv4EndPoint = new IPEndPoint(IPAddress.Parse("224.0.0.251"), 5353);
+        public static readonly IPEndPoint IPv6EndPoint = new IPEndPoint(IPAddress.Parse("ff02::fb"), 5353);
+
+        public static MdnsSocket CreateSocketIPv4(int index)
+        {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+
+            socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastInterface, IPAddress.HostToNetworkOrder(index));
+
+            socket.Bind(new IPEndPoint(IPAddress.Any, IPv4EndPoint.Port));
+            var tempEndpoint = new IPEndPoint(IPAddress.Any, IPv4EndPoint.Port);
+
+            IPAddress ip = IPv4EndPoint.Address;
+            socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.AddMembership, new MulticastOption(ip, index));
+            socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastTimeToLive, 1);
+            var info = new MdnsSocket(socket, IPv4EndPoint);
+            return info;
+        }
+
+        public static MdnsSocket CreateSocketIPv6(int index)
+        {
+            var socket = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp);
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+
+            // The MulticastInterface fails for IPv6, see: https://github.com/dotnet/runtime/issues/24255
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                LinuxHelper.MultiCastV6(socket, index);
+            }
+
+            socket.Bind(new IPEndPoint(IPAddress.IPv6Any, IPv6EndPoint.Port));
+            var tempEndpoint = new IPEndPoint(IPAddress.IPv6Any, IPv6EndPoint.Port);
+
+            IPAddress ip = IPv6EndPoint.Address;
+            socket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.AddMembership, new IPv6MulticastOption(ip, index));
+            socket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastTimeToLive, 1);
+            var info = new MdnsSocket(socket, IPv6EndPoint);
+            return info;
+        }
+
+        private MdnsSocket(Socket socket, IPEndPoint endpoint)
+        {
+            _socket = socket;
+            _endPoint = endpoint;
+            var tempEndpoint = new IPEndPoint(endpoint.Address, endpoint.Port);
+            _senderEndpoint = (EndPoint)tempEndpoint;
+        }
+
+        public delegate void OnReceivedFrom(IPAddress from, MemoryStream stream);
+        private Socket _socket;
+        private IPEndPoint _endPoint;
+        private EndPoint _senderEndpoint;
```

Comment:
Can you make `readonly` where possible.
And rename `_endPoint` to `_sendEndPoint`, and `_senderEndpoint` to `_receiveEndPoint`.

Reply:
Done